tests/migrations/test_commands.py

```diff
@@ -413,6 +414,59 @@ def test_regression_22823_unmigrated_fk_to_migrated_model(self):
         """
         call_command("migrate", "migrated_unapplied_app", stdout=six.StringIO())
 
+    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations_squashed"})
+    def test_migrate_record_replaced(self):
+        """
+        Running a single squashed migration should record all of the original
+        replaced migrations as run.
+        """
+        recorder = MigrationRecorder(connection)
+        out = six.StringIO()
+        call_command("migrate", "migrations", verbosity=0)
+        call_command("migrate", "migrations", list=True, stdout=out, no_color=True)
+        self.assertEqual(
+            'migrations\n'
+            ' [x] 0001_squashed_0002 (2 squashed migrations)\n',
+            out.getvalue().lower()
+        )
+        self.assertIn(
+            ("migrations", "0001_initial"),
+            recorder.applied_migrations()
+        )
+        self.assertIn(
+            ("migrations", "0002_second"),
+            recorder.applied_migrations()
+        )
+        self.assertIn(
+            ("migrations", "0001_squashed_0002"),
+            recorder.applied_migrations()
+        )
+        # Rollback changes
+        call_command("migrate", "migrations", "zero", verbosity=0)
+
+    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations_squashed"})
+    def test_migrate_record_squashed(self):
+        """
+        Regression test for #25231. Running migrate for a squashed migration
+        should record as run if all of the replaced migrations have been run.
+        """
+        recorder = MigrationRecorder(connection)
+        recorder.record_applied("migrations", "0001_initial")
+        recorder.record_applied("migrations", "0002_second")
+        out = six.StringIO()
+        call_command("migrate", "migrations", verbosity=0)
+        call_command("migrate", "migrations", list=True, stdout=out, no_color=True)
+        # No changes were actually applied so there is nothing to rollback
+        self.assertEqual(
```

Comment:
`squashmigrations` is only run by one person whereas the migration applied state needs to be updated in all instances of the project (dev environments, staging, prod, etc).

Here is where the loader considers them as applied https://github.com/django/django/blob/492537ac18df56c36f8a2335d773aa0fa01a61a3/django/db/migrations/loader.py#L253 Perhaps it should just be recorded then. That would mean the current `check_replacements` could be removed from the executor.